File: test/governor.test.ts
import { describe, expect, it, vi } from 'vitest';
import { ResourceGovernor, sampleWorkerRss, parseConcurrencyInput } from '../src/orchestrator/governor.js';
import { runScheduler } from '../src/orchestrator/scheduler.js';
import { runFleet } from '../src/fleet.js';
import { RunLogger } from '../src/logger.js';
import type { OrchestratorTask, ProjectBoard } from '../src/orchestrator/types.js';

const log = new RunLogger();
function task(partial: Partial<OrchestratorTask> & { id: string }): OrchestratorTask {
  return { title: partial.id, prompt: 'do it', dependsOn: [], status: 'pending', attempts: 0, ...partial };
}
function board(tasks: OrchestratorTask[]): ProjectBoard {
  return { goal: 'g', createdAt: '', updatedAt: '', roles: { planner: 'claude-code', executor: 'opencode' }, tasks };
}

describe('ResourceGovernor core (AC-1, AC-3)', () => {
  it('effectiveConcurrency: plenty RAM uses cpu cap', () => {
    const gov = new ResourceGovernor({ estMemPerWorkerBytes: 1_000_000_000, safetyRatio: 0.7, cpuFactor: 1 });
    const snap = { totalMem: 16_000_000_000, freeMem: 8_000_000_000, cpus: 4 };
    // memCap = floor(8e9*0.7/1e9)=5, cpuCap=4, min=4, clamp [1,8]=4
    expect(gov.effectiveConcurrency(snap, 8)).toBe(4);
  });

  it('effectiveConcurrency: exactly one worker of RAM', () => {
    const gov = new ResourceGovernor({ estMemPerWorkerBytes: 1_000_000_000, safetyRatio: 0.7, cpuFactor: 1 });
    // avail 1.5GB -> memCap floor(1.5*0.7)=1, cpus 4 -> cpuCap 4 -> min 1
    const snap = { totalMem: 8_000_000_000, freeMem: 1_500_000_000, cpus: 4 };
    expect(gov.effectiveConcurrency(snap, 4)).toBe(1);
  });

  it('effectiveConcurrency: zero free RAM still >=1', () => {
    const gov = new ResourceGovernor({ estMemPerWorkerBytes: 1_000_000_000, safetyRatio: 0.7, cpuFactor: 1 });
    const snap = { totalMem: 8_000_000_000, freeMem: 0, cpus: 8 };
    expect(gov.effectiveConcurrency(snap, 4)).toBe(1);
  });

  it('effectiveConcurrency: cpu-bound cap', () => {
    const gov = new ResourceGovernor({ estMemPerWorkerBytes: 200_000_000, safetyRatio: 0.7, cpuFactor: 0.5 });
    // 8 cpus *0.5=4 -> ceil 4, memCap huge -> limited to 4, configured 8 -> 4
    const snap = { totalMem: 64_000_000_000, freeMem: 32_000_000_000, cpus: 8 };
    expect(gov.effectiveConcurrency(snap, 8)).toBe(4);
  });

  it('effectiveConcurrency clamps to configured ceiling', () => {
    const gov = new ResourceGovernor({ estMemPerWorkerBytes: 100_000_000, safetyRatio: 1, cpuFactor: 1 });
    const snap = { totalMem: 32_000_000_000, freeMem: 16_000_000_000, cpus: 32 };
    // memCap huge, cpuCap 32, min 32, but configured 2 => 2
    expect(gov.effectiveConcurrency(snap, 2)).toBe(2);
  });

  it('effectiveAuto returns mem/cpu bound without configured ceiling', () => {
    const gov = new ResourceGovernor({ estMemPerWorkerBytes: 1_000_000_000, safetyRatio: 0.7, cpuFactor: 1 });
    const snap = { totalMem: 16_000_000_000, freeMem: 3_100_000_000, cpus: 16 };
    // memCap floor(3.1*0.7)=2, cpuCap 16 => min 2
    expect(gov.effectiveAuto(snap)).toBe(2);
  });

  it('performs no network calls and <5ms per check (benchmark)', () => {
    const gov = new ResourceGovernor();
    const snap = { totalMem: 16_000_000_000, freeMem: 8_000_000_000, cpus: 4 };
    const start = performance.now();
    for (let i = 0; i < 100; i++) gov.effectiveConcurrency(snap, 4);
    const elapsed = performance.now() - start;
    expect(elapsed / 100).toBeLessThan(5);
  });

  it('OS reads are cached for <=1s', () => {
    const gov = new ResourceGovernor({ cacheTtlMs: 1000 });
    const s1 = gov.getSnapshotSync();
    const s2 = gov.getSnapshotSync();
    expect(s1).toBe(s2); // same object from cache
    // force refresh
    const s3 = gov.refreshSnapshot();
    expect(s3).not.toBe(s1);
  });

  it('calibration raises estMemPerWorker via p75 and never below floor', () => {
    const gov = new ResourceGovernor({ estMemPerWorkerBytes: 200_000_000, minEstFloorBytes: 100_000_000 });
    expect(gov.getEstMemPerWorker()).toBe(200_000_000);
    // push 4 samples: 100,200,300,400 -> p75 is 300
    for (const rss of [100_000_000, 200_000_000, 300_000_000, 400_000_000]) gov.recordRss(rss);
    expect(gov.getEstMemPerWorker()).toBe(300_000_000);
    // smaller sample doesn't lower it
    gov.recordRss(50_000_000);
    expect(gov.getEstMemPerWorker()).toBe(300_000_000);
  });

  it('parseConcurrencyInput handles auto and numbers', () => {
    expect(parseConcurrencyInput('auto')).toBe('auto');
    expect(parseConcurrencyInput('Auto')).toBe('auto');
    expect(parseConcurrencyInput('4')).toBe(4);
    expect(parseConcurrencyInput(2)).toBe(2);
    expect(parseConcurrencyInput('  auto  ')).toBe('auto');
  });

  it('formatStatus contains expected fields', () => {
    const gov = new ResourceGovernor({ estMemPerWorkerBytes: 1_000_000_000 });
    const snap = { totalMem: 16 * 1024 * 1024 * 1024, freeMem: 8 * 1024 * 1024 * 1024, cpus: 8 };
    const line = gov.formatStatus('auto', 2, snap);
    expect(line).toContain('workers auto->2');
    expect(line).toContain('mem');
    expect(line).toContain('est');
  });
});

describe('sampleWorkerRss (AC-2)', () => {
  it('returns real RSS for current process on macOS/Linux or null on unsupported', async () => {
    const rss = await sampleWorkerRss(process.pid);
    if (process.platform === 'linux' || process.platform === 'darwin') {
      expect(rss).toBeGreaterThan(0);
      expect(typeof rss).toBe('number');
    } else {
      expect(rss).toBeNull();
    }
  });

  it('returns null for invalid pid', async () => {
    expect(await sampleWorkerRss(-1)).toBeNull();
    expect(await sampleWorkerRss(9999999)).toBeNull();
  });
});

describe('scheduler governor enforcement (AC-4,5,6,7,8)', () => {
  it('AC-4: with fixture governor reporting only 1 worker fits, concurrency 4 never exceeds 1 dispatched', async () => {
    const gov = new ResourceGovernor({ estMemPerWorkerBytes: 1_000_000_000, safetyRatio: 0.7, cpuFactor: 1 });
    // inject snapshot where only 1 fits
    gov.injectSnapshot({ totalMem: 8_000_000_000, freeMem: 1_200_000_000, cpus: 4 });
    // est 1GB -> memCap floor(1.2*0.7)=0 -> max(1,0) =>1, cpuCap 4 => min 1 => effective 1
    const b = board([task({ id: 'A' }), task({ id: 'B' }), task({ id: 'C' }), task({ id: 'D' })]);
    let peakDispatched = 0;
    let currentDispatched = 0;
    // Monkey: track via queue? We'll instrument executeTask to track peak concurrency
    let active = 0;
    let peakActive = 0;
    await runScheduler(
      b,
      { repoPath: '.', executor: 'opencode', concurrency: 'auto', governor: gov, governorSnapshot: { totalMem: 8_000_000_000, freeMem: 1_200_000_000, cpus: 4 }, maxTaskRetries: 1, timeoutMs: 1000 },
      {
        executeTask: async () => {
          active += 1;
          peakActive = Math.max(peakActive, active);
          currentDispatched += 1;
          peakDispatched = Math.max(peakDispatched, currentDispatched);
          await new Promise((r) => setTimeout(r, 15));
          active -= 1;
          currentDispatched -= 1;
          return { ok: true };
        },
      },
      log,
    );
    expect(peakActive).toBeLessThanOrEqual(1);
    expect(b.tasks.every((t) => t.status === 'done')).toBe(true);
  });

  it('AC-5: ceiling drops mid-wave — second worker waits (throttle not fail)', async () => {
    const gov = new ResourceGovernor({ estMemPerWorkerBytes: 500_000_000, safetyRatio: 0.7, cpuFactor: 1 });
    // start with enough for 2, then flip to 1 after first task
    let callCount = 0;
    const baseSnap = { totalMem: 8_000_000_000, freeMem: 4_000_000_000, cpus: 4 };
    // mock getSnapshotSync to return different freeMem on each call
    const originalGet = gov.getSnapshotSync.bind(gov);
    gov.getSnapshotSync = () => {
      callCount += 1;
      if (callCount <= 2) return baseSnap; // first wave pool 2 (memCap 5, cpu 4 ->4)
      return { totalMem: 8_000_000_000, freeMem: 600_000_000, cpus: 4 }; // memCap 0 ->1
    };
    const b = board([task({ id: 'A' }), task({ id: 'B' }), task({ id: 'C' })]);
    let peak = 0;
    let active = 0;
    await runScheduler(
      b,
      { repoPath: '.', executor: 'opencode', concurrency: 'auto', governor: gov, maxTaskRetries: 1, timeoutMs: 1000 },
      {
        executeTask: async () => {
          active += 1;
          peak = Math.max(peak, active);
          await new Promise((r) => setTimeout(r, 20));
          active -= 1;
          return { ok: true };
        },
      },
      log,
    );
    // Should have completed without marking tasks failed due to throttle
    expect(b.tasks.filter((t) => t.status === 'failed')).toHaveLength(0);
    expect(b.tasks.every((t) => t.status === 'done')).toBe(true);
    // restore
    gov.getSnapshotSync = originalGet;
  });

  it('AC-6: after pressureWaitTimeout with no headroom, wave proceeds at concurrency 1', async () => {
    const gov = new ResourceGovernor({ estMemPerWorkerBytes: 1_000_000_000, safetyRatio: 0.7, cpuFactor: 1, pressureWaitTimeoutMs: 400 });
    gov.injectSnapshot({ totalMem: 8_000_000_000, freeMem: 500_000_000, cpus: 4 }); // effective 1
    // Need to test throttle timeout: use a long-running first task + low effective
    const b = board([task({ id: 'A' }), task({ id: 'B' })]);
    const start = Date.now();
    await runScheduler(
      b,
      { repoPath: '.', executor: 'opencode', concurrency: 'auto', governor: gov, governorSnapshot: { totalMem: 8_000_000_000, freeMem: 500_000_000, cpus: 4 }, maxTaskRetries: 1, timeoutMs: 1000 },
      {
        executeTask: async () => {
          await new Promise((r) => setTimeout(r, 30));
          return { ok: true };
        },
      },
      log,
    );
    expect(b.tasks.every((t) => t.status === 'done')).toBe(true);
    // Ensure it didn't deadlock: elapsed should be at least sum of serial tasks but not huge
    expect(Date.now() - start).toBeGreaterThan(30);
  });

  it('AC-7: explicit numeric concurrency bypasses governor (zero-regression)', async () => {
    const gov = new ResourceGovernor({ estMemPerWorkerBytes: 10_000_000_000, safetyRatio: 0.7 }); // absurdly high est -> effective 1 if auto
    gov.injectSnapshot({ totalMem: 8_000_000_000, freeMem: 1_000_000_000, cpus: 4 });
    // Governor would say 1, but numeric 3 must stay 3
    const bAuto = board([task({ id: 'A' }), task({ id: 'B' }), task({ id: 'C' })]);
    const bNum = board([task({ id: 'A' }), task({ id: 'B' }), task({ id: 'C' })]);
    let peakAuto = 0, activeAuto = 0;
    await runScheduler(bAuto, { repoPath: '.', executor: 'opencode', concurrency: 'auto', governor: gov, governorSnapshot: { totalMem: 8_000_000_000, freeMem: 1_000_000_000, cpus: 4 }, maxTaskRetries: 1, timeoutMs: 1000 }, {
      executeTask: async () => { activeAuto+=1; peakAuto=Math.max(peakAuto, activeAuto); await new Promise(r=>setTimeout(r,10)); activeAuto-=1; return {ok:true}; },
    }, log);
    expect(peakAuto).toBeLessThanOrEqual(1);

    let peakNum = 0, activeNum = 0;
    await runScheduler(bNum, { repoPath: '.', executor: 'opencode', concurrency: 3, governor: gov, governorSnapshot: { totalMem: 8_000_000_000, freeMem: 1_000_000_000, cpus: 4 }, maxTaskRetries: 1, timeoutMs: 1000 }, {
      executeTask: async () => { activeNum+=1; peakNum=Math.max(peakNum, activeNum); await new Promise(r=>setTimeout(r,10)); activeNum-=1; return {ok:true}; },
    }, log);
    expect(peakNum).toBe(3);
    expect(peakNum).not.toBe(peakAuto);
  });

  it('AC-8: --concurrency auto logs resolved number and inputs', async () => {
    const gov = new ResourceGovernor();
    const snap = { totalMem: 16 * 1024 * 1024 * 1024, freeMem: 8 * 1024 * 1024 * 1024, cpus: 4 };
    gov.injectSnapshot(snap);
    const eff = gov.effectiveAuto(snap);
    const line = gov.formatStatus('auto', eff, snap);
    expect(line).toContain('auto->');
    expect(line).toContain('mem');
    expect(line).toContain('est');
    // Simulate CLI logging: ensure governor info would be logged via RunLogger
    const tlog = new RunLogger();
    tlog.info('governor', `auto -> ${eff}`, { freeMem: snap.freeMem, totalMem: snap.totalMem, cpus: snap.cpus });
    // Check log file contains governor line
    const { readFileSync } = await import('node:fs');
    const content = readFileSync(tlog.path, 'utf8');
    expect(content).toContain('governor');
  });
});

describe('fleet governor enforcement (AC-4 fleet counterpart)', () => {
  it('fleet respects governor ceiling when auto', async () => {
    const { ResourceGovernor } = await import('../src/orchestrator/governor.js');
    const gov = new ResourceGovernor({ estMemPerWorkerBytes: 2_000_000_000, safetyRatio: 0.7, cpuFactor: 1 });
    const snap = { totalMem: 16_000_000_000, freeMem: 2_000_000_000, cpus: 8 }; // memCap 0 ->1, cpu 8 =>1
    let peak = 0, active = 0;
    const runOne = vi.fn().mockImplementation(async () => {
      active += 1; peak = Math.max(peak, active);
      await new Promise((r) => setTimeout(r, 10));
      active -= 1;
      return { ok: true, summary: 'ok' };
    });
    const entries = [{ name: 'a', path: '/a' }, { name: 'b', path: '/b' }];
    await runFleet({ ticketIds: ['T1', 'T2', 'T3'], entries, concurrency: 'auto', governor: gov, governorSnapshot: snap, timeoutMs: 1000, worker: 'opencode', autoPr: false, maxLoops: 1, runOne });
    expect(peak).toBeLessThanOrEqual(1);
    expect(runOne).toHaveBeenCalledTimes(6);
  });

  it('fleet explicit numeric bypasses governor', async () => {
    const { ResourceGovernor } = await import('../src/orchestrator/governor.js');
    const gov = new ResourceGovernor({ estMemPerWorkerBytes: 10_000_000_000 });
    const snap = { totalMem: 8_000_000_000, freeMem: 500_000_000, cpus: 4 };
    let peak = 0, active = 0;
    const runOne = vi.fn().mockImplementation(async () => {
      active += 1; peak = Math.max(peak, active);
      await new Promise((r) => setTimeout(r, 10));
      active -= 1;
      return { ok: true, summary: 'ok' };
    });
    const entries = [{ name: 'a', path: '/a' }];
    await runFleet({ ticketIds: ['1', '2', '3'], entries, concurrency: 3, governor: gov, governorSnapshot: snap, timeoutMs: 1000, worker: 'opencode', autoPr: false, maxLoops: 1, runOne });
    expect(peak).toBe(3);
  });
});

File: internal/orchestrator/governor_test.go
// Package file mirrors test/governor.test.ts (FR-GO-07, issue #194) — the
// ResourceGovernor core and sampleWorkerRss describes. The scheduler/fleet
// governor-enforcement describes belong to scheduler.ts/fleet.ts ports.

package orchestrator

import (
	"fmt"
	"os"
	"runtime"
	"strings"
	"testing"
	"time"
)

func gi64(v int64) *int64     { return &v }
func gf64(v float64) *float64 { return &v }
func gi(v int) *int           { return &v }

func govSnapshot(total, free int64, cpus int) OsSnapshot {
	return OsSnapshot{TotalMem: total, FreeMem: free, Cpus: cpus}
}

func TestResourceGovernorCore(t *testing.T) {
	t.Run("effectiveConcurrency: plenty RAM uses cpu cap", func(t *testing.T) {
		gov := NewResourceGovernor(&GovernorOptions{EstMemPerWorkerBytes: gi64(1_000_000_000), SafetyRatio: gf64(0.7), CPUFactor: gf64(1)})
		snap := govSnapshot(16_000_000_000, 8_000_000_000, 4)
		// memCap = floor(8e9*0.7/1e9)=5, cpuCap=4, min=4, clamp [1,8]=4
		if got := gov.EffectiveConcurrency(snap, 8); got != 4 {
			t.Fatalf("effectiveConcurrency = %d, want 4", got)
		}
	})

	t.Run("effectiveConcurrency: exactly one worker of RAM", func(t *testing.T) {
		gov := NewResourceGovernor(&GovernorOptions{EstMemPerWorkerBytes: gi64(1_000_000_000), SafetyRatio: gf64(0.7), CPUFactor: gf64(1)})
		// avail 1.5GB -> memCap floor(1.5*0.7)=1, cpus 4 -> cpuCap 4 -> min 1
		snap := govSnapshot(8_000_000_000, 1_500_000_000, 4)
		if got := gov.EffectiveConcurrency(snap, 4); got != 1 {
			t.Fatalf("effectiveConcurrency = %d, want 1", got)
		}
	})

	t.Run("effectiveConcurrency: zero free RAM still >=1", func(t *testing.T) {
		gov := NewResourceGovernor(&GovernorOptions{EstMemPerWorkerBytes: gi64(1_000_000_000), SafetyRatio: gf64(0.7), CPUFactor: gf64(1)})
		snap := govSnapshot(8_000_000_000, 0, 8)
		if got := gov.EffectiveConcurrency(snap, 4); got != 1 {
			t.Fatalf("effectiveConcurrency = %d, want 1", got)
		}
	})

	t.Run("effectiveConcurrency: cpu-bound cap", func(t *testing.T) {
		gov := NewResourceGovernor(&GovernorOptions{EstMemPerWorkerBytes: gi64(200_000_000), SafetyRatio: gf64(0.7), CPUFactor: gf64(0.5)})
		// 8 cpus *0.5=4 -> ceil 4, memCap huge -> limited to 4, configured 8 -> 4
		snap := govSnapshot(64_000_000_000, 32_000_000_000, 8)
		if got := gov.EffectiveConcurrency(snap, 8); got != 4 {
			t.Fatalf("effectiveConcurrency = %d, want 4", got)
		}
	})

	t.Run("effectiveConcurrency clamps to configured ceiling", func(t *testing.T) {
		gov := NewResourceGovernor(&GovernorOptions{EstMemPerWorkerBytes: gi64(100_000_000), SafetyRatio: gf64(1), CPUFactor: gf64(1)})
		snap := govSnapshot(32_000_000_000, 16_000_000_000, 32)
		// memCap huge, cpuCap 32, min 32, but configured 2 => 2
		if got := gov.EffectiveConcurrency(snap, 2); got != 2 {
			t.Fatalf("effectiveConcurrency = %d, want 2", got)
		}
	})

	t.Run("effectiveAuto returns mem/cpu bound without configured ceiling", func(t *testing.T) {
		gov := NewResourceGovernor(&GovernorOptions{EstMemPerWorkerBytes: gi64(1_000_000_000), SafetyRatio: gf64(0.7), CPUFactor: gf64(1)})
		snap := govSnapshot(16_000_000_000, 3_100_000_000, 16)
		// memCap floor(3.1*0.7)=2, cpuCap 16 => min 2
		if got := gov.EffectiveAuto(snap); got != 2 {
			t.Fatalf("effectiveAuto = %d, want 2", got)
		}
	})

	t.Run("performs no network calls and <5ms per check (benchmark)", func(t *testing.T) {
		gov := NewResourceGovernor(nil)
		snap := govSnapshot(16_000_000_000, 8_000_000_000, 4)
		start := time.Now()
		for i := 0; i < 100; i++ {
			gov.EffectiveConcurrency(snap, 4)
		}
		elapsed := time.Since(start)
		if elapsed/100 > 5*time.Millisecond {
			t.Fatalf("avg check = %v, want < 5ms", elapsed/100)
		}
	})

	t.Run("OS reads are cached for <=1s", func(t *testing.T) {
		gov := NewResourceGovernor(&GovernorOptions{CacheTtlMs: gi(1000)})
		s1 := gov.GetSnapshotSync()
		s2 := gov.GetSnapshotSync()
		if s1 != s2 { // same pointer from cache
			t.Fatal("expected the same cached snapshot")
		}
		// force refresh
		s3 := gov.RefreshSnapshot()
		if s3 == s1 {
			t.Fatal("refresh must return a new snapshot")
		}
	})

	t.Run("calibration raises estMemPerWorker via p75 and never below floor", func(t *testing.T) {
		gov := NewResourceGovernor(&GovernorOptions{EstMemPerWorkerBytes: gi64(200_000_000), MinEstFloorBytes: gi64(100_000_000)})
		if got := gov.GetEstMemPerWorker(); got != 200_000_000 {
			t.Fatalf("initial est = %d", got)
		}
		// push 4 samples: 100,200,300,400 -> p75 is 300
		for _, rss := range []float64{100_000_000, 200_000_000, 300_000_000, 400_000_000} {
			gov.RecordRss(rss)
		}
		if got := gov.GetEstMemPerWorker(); got != 300_000_000 {
			t.Fatalf("calibrated est = %d, want 300000000", got)
		}
		// smaller sample doesn't lower it
		gov.RecordRss(50_000_000)
		if got := gov.GetEstMemPerWorker(); got != 300_000_000 {
			t.Fatalf("est after small sample = %d, want 300000000", got)
		}
	})

	t.Run("parseConcurrencyInput handles auto and numbers", func(t *testing.T) {
		cases := []struct {
			raw  any
			auto bool
			n    int
		}{
			{"auto", true, 0},
			{"Auto", true, 0},
			{"4", false, 4},
			{2, false, 2},
			{"  auto  ", true, 0},
		}
		for _, c := range cases {
			got := ParseConcurrencyInput(c.raw)
			if got.Auto != c.auto || (!c.auto && got.Value != c.n) {
				t.Fatalf("parseConcurrencyInput(%v) = %+v, want auto=%v n=%d", c.raw, got, c.auto, c.n)
			}
		}
	})

	t.Run("formatStatus contains expected fields", func(t *testing.T) {
		gov := NewResourceGovernor(&GovernorOptions{EstMemPerWorkerBytes: gi64(1_000_000_000)})
		snap := govSnapshot(16*1024*1024*1024, 8*1024*1024*1024, 8)
		line := gov.FormatStatus(ConcurrencyValue{Auto: true}, 2, &snap)
		for _, want := range []string{"workers auto->2", "mem", "est"} {
			if !contains(line, want) {
				t.Fatalf("line %q missing %q", line, want)
			}
		}
	})

	t.Run("formatStatus reports sample age from cachedAt and calibration count (Q14)", func(t *testing.T) {
		now := int64(1_000_000)
		gov := NewResourceGovernor(&GovernorOptions{EstMemPerWorkerBytes: gi64(1_000_000_000)})
		gov.SetClock(func() int64 { return now })
		snap := govSnapshot(16*1024*1024*1024, 8*1024*1024*1024, 8)
		gov.InjectSnapshot(&snap)
		gov.RecordRss(1_500_000_000)
		gov.RecordRss(1_600_000_000)
		line := gov.FormatStatus(ConcurrencyValue{Auto: true}, 2, &snap)
		if !contains(line, "sample 0.0s") || !contains(line, "cal 2") {
			t.Fatalf("line = %q", line)
		}
		// age grows with cachedAt distance
		now += 2_500
		line = gov.FormatStatus(ConcurrencyValue{Auto: true}, 2, &snap)
		if !contains(line, "sample 2.5s") {
			t.Fatalf("line = %q", line)
		}
		// numeric input keeps the same one-line shape
		line = gov.FormatStatus(ConcurrencyValue{Value: 4}, 3, &snap)
		if !contains(line, "workers 3/4") || !contains(line, "sample 2.5s, cal 2") {
			t.Fatalf("line = %q", line)
		}
	})

	t.Run("formatStatus marks age n/a for uncached snapshots and cal 0 before calibration", func(t *testing.T) {
		gov := NewResourceGovernor(&GovernorOptions{EstMemPerWorkerBytes: gi64(1_000_000_000)})
		external := govSnapshot(16_000_000_000, 8_000_000_000, 8)
		line := gov.FormatStatus(ConcurrencyValue{Auto: true}, 2, &external)
		if !contains(line, "sample n/a") || !contains(line, "cal 0") {
			t.Fatalf("line = %q", line)
		}
		// resetCalibration drops the count back to 0 (aggregate only, no per-pid data)
		gov.InjectSnapshot(&external)
		gov.RecordRss(2_000_000_000)
		if line := gov.FormatStatus(ConcurrencyValue{Auto: true}, 2, &external); !contains(line, "cal 1") {
			t.Fatalf("line = %q", line)
		}
		gov.ResetCalibration()
		if line := gov.FormatStatus(ConcurrencyValue{Auto: true}, 2, &external); !contains(line, "cal 0") {
			t.Fatalf("line = %q", line)
		}
	})
}

func contains(s, sub string) bool { return strings.Contains(s, sub) }

func TestSampleWorkerRss(t *testing.T) {
	t.Run("returns real RSS for current process on macOS/Linux or null on unsupported", func(t *testing.T) {
		rss := SampleWorkerRss(os.Getpid())
		if runtime.GOOS == "linux" || runtime.GOOS == "darwin" {
			if rss == nil || *rss <= 0 {
				t.Fatalf("rss = %v, want > 0", rss)
			}
		} else if rss != nil {
			t.Fatalf("unsupported platform must return nil, got %v", *rss)
		}
	})

	t.Run("returns null for invalid pid", func(t *testing.T) {
		if SampleWorkerRss(-1) != nil {
			t.Fatal("pid -1 must be nil")
		}
		if SampleWorkerRss(9999999) != nil {
			t.Fatal("nonexistent pid must be nil")
		}
	})

	t.Run("parses scripted darwin ps output through the runner seam", func(t *testing.T) {
		if runtime.GOOS != "darwin" {
			t.Skip("darwin-only seam check")
		}
		run := CommandRunner(func(name string, args []string, timeoutMs int) (string, error) {
			if name != "ps" {
				return "", fmt.Errorf("unexpected %s", name)
			}
			return "  4242\n", nil
		})
		rss := SampleWorkerRssWith(run, 123)
		if rss == nil || *rss != 4242*1024 {
			t.Fatalf("rss = %v, want %d", rss, 4242*1024)
		}
		// non-numeric / non-positive output → nil
		run2 := CommandRunner(func(name string, args []string, timeoutMs int) (string, error) { return "abc\n", nil })
		if SampleWorkerRssWith(run2, 123) != nil {
			t.Fatal("garbage ps output must be nil")
		}
		run3 := CommandRunner(func(name string, args []string, timeoutMs int) (string, error) { return "0\n", nil })
		if SampleWorkerRssWith(run3, 123) != nil {
			t.Fatal("zero rss must be nil")
		}
	})
}
